### ISSUE TITLE
Replacement feature

### ISSUE THREAD

**Issue by popouille** [matchory/response-cache#3]
Hello,

This is possible to add a replacement feature to handle Csrf tokan and dynamic content such content depending of cookies?

**Comment by Radiergummi** [comment]
Hi @popouille, yes, this is possible, although I wouldn't recommend it. The package allows adding your own, custom caching strategy. Take a look at the ["Customize Tagging" section of the Readme](https://github.com/matchory/response-cache#customizing-tagging) for an example of a custom strategy.  
Basically, a strategy allows you to customize how the cache key is generated, and which tags a response will be stored with. If you wanted to check the value of a cookie before caching the response, or append the cookie value to the cache key (thus creating a unique cached variant for every possible cookie value), you could use a strategy like the following:
```php
use Matchory\ResponseCache\Support\BaseStrategy;

class Strategy extends BaseStrategy
{
    /**
     * @inheritDoc
     */
    protected function buildSuffix(Request $request): string
    {
        // Fetch the user ID from the request. This is what the base strategy does to
        // scope cached responses to a single user. 
        $userId = $this->auth->check()
            ? (string)$this->auth->id()
            : '';

        // Fetch the value of your custom cookie, or CSRF token, or whatever from the
        // request instance. It's the same thing you'd get inside a controller method.
        $customCookieValue = $request->cookie('custom-cookie-name');

        // Remember that the result of this method will be hashed, so you don't need
        // any "human-readable" separator characters or something.
        return $userId . $customCookieValue;
    }
}
```
Strategies support a range of other useful customization options, just check out the [Base Strategy](https://github.com/matchory/response-cache/blob/master/src/Support/BaseStrategy.php) to see what else is possible.

Once your strategy is done, register it in your `app/Providers/AppServiceProvider`:
```php
    public function register(): void
    {
        // ...
        $this->app->bind(CacheStrategy::class, Strategy::class);
    }
```

Let me know if you need further help ✌️

**Comment by popouille** [comment]
Yes that why i use your package instead of spatie one, because we can set custom key and tag outside of route.php. But for some scenarios isnt enought;
I have a cookie for last visited books thats store 6 books ids from 15K books.
So its a lot of id possiblities, if i set this cookies content as a part of the key, each time a guest will visit a page, a new cache item will be generated, thats means a lot of differents cached items...
But with a replacement feature, only one item is cached, and display to all guests, and the page content is replaced depending of each guest cookie content.

**Comment by Radiergummi** [comment]
I'm not exactly sure what this feature is supposed to do. The response cache stores a response object depending on some request parameters, and can be invalidated independently by one or more tags. Could you provide a little more detail what you want to see implemented here?

**Comment by popouille** [comment]
I have 15000 books informations pages on my website.
When a guest visit one of these page it stores the page url and book id in a cookie "recently views".


If i cache these page, then to get the widget base on each guest "recently views" cookie, i have to put this cookie content as part of the custom cache key, so the key will contains for example:
`id500.id103,id5002.id5410`
but i can also be: 
`id20,id500,id910,id410`. etc etc
But if i do it, each time a guest will visit a new book page, it will store it as a new key in the cache (because there is little chance that different visitors visit the same pages and even less in the same order). So for an heavy traffic website, its maybe one new page store in cache each second, it will lead to a huge cache size. 
To summarize if i use the cache key to handle this: for one book page, i will have multiple page variation in the cache.

If i use a replacer feature. Only 1 page per book will be cached, then i dynamically replace the widget content depending of the "recently views" cookie content of each Guest. I get the dynamic content before response is cached and replace it by a tag, then i replace the tag in the cached response by my dynamic content.

**Comment by popouille** [comment]
I did this, isnt perfect but its a start:

`<?php

namespace Matchory\ResponseCache\Replacers;

use Symfony\Component\HttpFoundation\Response;

class HistoryReplacer implements Replacer
{
    public function prepareResponseToCache(Response $response): void
    {
        if (! $response->getContent()) {
            return;
        }

        $recently_viewed_content = json_decode(\Cookie::get('recently_viewed'), true);
        if ($recently_viewed_content == null) {
            $recently_viewed_content = [];
        }
        if ($recently_viewed_content) {
            krsort($recently_viewed_content);
        }

        $historyData = [];
        $replacementString = [];
        $count = 0;
        foreach ($recently_viewed_content as $rvc) {
            if (request()->url() !== $rvc['url']) {
                $count++;
                array_push($historyData, $rvc['url'], $rvc['name'], $rvc['cover'], $rvc['visited_at']);
                array_push($replacementString, '<laravel-responsecache-rdc-url-'.$count.'>', '<laravel-responsecache-rdc-name-'.$count.'>', '<laravel-responsecache-rdc-cover-'.$count.'>', '<laravel-responsecache-rdc-visited-'.$count.'>');
            }
        }

        $response->setContent(str_replace($historyData, $replacementString, $response->getContent()));
    }

    public function replaceInCachedResponse(Response $response): void
    {
        if (! $response->getContent()) {
            return;
        }

        $recently_viewed_content = json_decode(\Cookie::get('recently_viewed'), true);
        if ($recently_viewed_content == null) {
            $recently_viewed_content = [];
        }
        if ($recently_viewed_content) {
            krsort($recently_viewed_content);
        }

        $historyData = [];
        $replacementString = [];
        $count = 0;
        foreach ($recently_viewed_content as $rvc) {
            if (request()->url() !== $rvc['url']) {
                $count++;
                array_push($historyData, $rvc['url'], $rvc['name'], $rvc['cover'], $rvc['visited_at']);
                array_push($replacementString, '<laravel-responsecache-rdc-url-'.$count.'>', '<laravel-responsecache-rdc-name-'.$count.'>', '<laravel-responsecache-rdc-cover-'.$count.'>', '<laravel-responsecache-rdc-visited-'.$count.'>');
            }
        }


        $response->setContent(str_replace(
            $replacementString,
            $historyData,
            $response->getContent()
        ));
    }
}`

**Comment by Radiergummi** [comment]
Wouldn't it be easier and faster to fetch the recommendations via AJAX from the client? You could even cache the recommendations endpoint for the current set of books, so if someone visits a book page, then navigates away, and visits another book page later on, they'd be served the cached copy of their recommendations.

**Comment by Radiergummi** [comment]
Other than that, maybe caching the full response just isn't what you want to do, opposed to something like this:
```php
public function booksRoute(Request $request, string $bookId): Response
{
    $pageContent = \Cache::remember(
        key: 'book_page_' . $bookId,
        ttl: 3600,
        callback: fn(): string => $this->renderBookPage()
    );

    $recentlyViewed = \json_decode($request->cookie('recently_viewed'), true);   ​

    // do replacement stuff with recently viewed data

    $responseBody = \str_replace(
        $replacementString,
        $historyData,
        $pageContent
    );

    return \response($responseBody, 200);
}
```

By using [`Cache::remember()`](https://laravel.com/docs/8.x/cache#retrieve-store), you can dynamically cache any page that doesn't exist in the cache yet, then replace the dynamic content afterwards. This should be very fast from a performance point of view, and maybe is easier than implementing the whole infrastructure for replacing content in cached responses.

**Comment by popouille** [comment]
Perfect thank you very much

**Comment by Radiergummi** [comment]
Glad you got it working!
